src/Sulu/Component/HttpCache/EventSubscriber/InvalidationSubscriber.php

```diff
@@ -57,25 +91,100 @@ public function __construct(
 
-    public function invalidateDocumentForPublishing(PublishEvent $event)
+    /**
+     * Invalidates all urls in in the locale of the document of already published documents when they
+     * get republished (eg on content change).
+     * This method is executed before the actual publishing of the document.
+     *
+     * @param PublishEvent $event
+     */
+    public function invalidateDocumentBeforePublishing(PublishEvent $event)
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof ResourceSegmentBehavior || !$document->getPublished()) {
+            // no need to invalidate the path to the document if it is not published currently
+            return;
+        }
+
+        // invalidate urls of the locale of the document
+        $locale = $this->documentInspector->getLocale($document);
+        foreach ($this->getLocaleUrls($document, $locale) as $url) {
+            $this->pathHandler->invalidatePath($url);
+        }
+    }
+
+    /**
+     * Invalidates all urls in all locales of the document whey they get removed.
+     * This method is executed before the actual removing of the document.
+     *
+     * @param RemoveEvent $event
+     */
+    public function invalidateDocumentBeforeRemoving(RemoveEvent $event)
     {
         $document = $event->getDocument();
 
-        if (!$document instanceof StructureBehavior) {
+        if (!$document instanceof ResourceSegmentBehavior) {
             return;
         }
 
-        $structureBridge = $this->structureManager->wrapStructure(
-            $this->documentInspector->getMetadata($document)->getAlias(),
-            $this->documentInspector->getStructureMetadata($document)
-        );
+        // invalidate urls of all locales of the document
+        $urls = [];
+        foreach ($this->documentInspector->getLocales($document) as $locale) {
+            $urls = array_merge($urls, $this->getLocaleUrls($document, $locale));
+        }
+
+        foreach ($urls as $url) {
+            $this->pathHandler->invalidatePath($url);
+        }
+    }
 
-        $structureBridge->setDocument($document);
+    /**
+     * Returns all urls of the given locale which are associated with the given document.
+     * The returned array contains all current urls and all history urls.
+     * The returned urls can contain placeholders (eg {host}).
+     *
+     * @param ResourceSegmentBehavior $document
+     * @param $locale
+     *
+     * @return array Urls of the given locale which are associated with the given document
+     */
+    private function getLocaleUrls(ResourceSegmentBehavior $document, $locale)
+    {
+        $documentUuid = ($document instanceof UuidBehavior) ? $document->getUuid() : null;
+        $webspace = ($document instanceof WebspaceBehavior) ? $document->getWebspaceName() : null;
+
+        if (!$document || !$locale || !$documentUuid) {
+            return [];
+        }
+
+        // get current rlp and history-rlps
+        $rlps = [];
```

Comment:
Please try to avoid the term `rlp`. I don't know how that slipped in, but `rlp` stands for resource locator part, whereby it never is only a part of the resource locator. Apart from that we don't use abbreviations. Just call it `resourceLocators`.